### ISSUE TITLE
Add <form> (or array of exceptions) to list of valid elements for no-static-element-interactions. 

### ISSUE THREAD

**Issue by jaepage** [evcohen/eslint-plugin-jsx-a11y#153]
I wish to detect with a meta key is pressed with the return/enter key upon form submission via keyboard navigation: 

```
form.addEventListener('keydown', function (event) {
  console.log('keydown', event.target.name, event.metaKey);
});
```

However, this plugin warns of violation of [`no-static-element-interactions` rule](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-static-element-interactions.md). 

This seems to suggest the event listener should be added to every single interactive element on the form, which would be tedious and would present conflicts with other listeners. 

Could `<form>` be added to the list of interactive elements? Could an array of exceptions be provided as an option to the linter's config?

**Comment by evcohen** [comment]
This should be:

```
<form onSubmit={whateverYourSubmitFnIs}>
   ....
   <button type="submit" />
</form>
```

Then in your onSubmit fn, you do `e.preventDefault()`

I think this will pick up on the enter button for submit

**Comment by jaepage** [comment]
I need to capture the keydown before the submit. 

My form has two submit buttons. If a user hits <kbd>return</kbd> anywhere on the form (outside `<textarea>` elements), the first submit button in the DOM tree is triggered. 

This event listener provides the option for users to trigger the second submit button via <kbd>command</kbd><kbd>return</kbd>.

**Comment by ljharb** [comment]
@jaepage a user hitting `return` doesn't necessarily trigger *any* submit button - in fact, it's generally not reliable and a bad idea to ever have more than one submit button, because it's generally not reliable to determine which one was clicked.

**Comment by jaepage** [comment]
No need to at-ping me, please. I'm subscribed to the thread. 

Our form has two buttons that each submit to the API with a hidden flag turned on or off (we actually previously had the two buttons submitting to two different APIs). 

It's valid HTML to have a form with two submit buttons, per https://html5.validator.nu/ and https://validator.w3.org/. 

By adding such a listener to the form, we could prevent submission of all hits of enter/return, send command-returns to one button, and shift-command-returns to the other, etc.

**Comment by ljharb** [comment]
(O.o if you're subscribed, then pinging you won't give you a different notification, and if you're not, then this will resubscribe you - it's exceedingly common github etiquette to at-mention people when you want to be sure your comment is seen as a direct reply)

It may be valid HTML, but all browsers don't implement that spec consistently. The proper solution would be to have one submit button, and a checkbox for the flag, and then to use JS to map the checkbox so it looks like a button, checks the box on click, and then auto-submits the form.

That said, if you wanted to listen to `shift-return`, you'd certainly need a key listener on the `<form>`, since as you pointed out, it'd be impractical to put the listener on every element.

I think adding an option to the rule for an explicit list of exceptions (ie, `form`, `MyComponent`, etc) makes perfect sense here.

**Comment by evcohen** [comment]
@ljharb (sorry for the double notification)...I think I'm against this config option, but can be convinced. I think users should manually disable the rule via comments in these cases. This indicates in the source that this _may_ be bad design for a11y reasons. By obscuring the ignored components into an .eslintrc file, users may be unknowingly implementing inaccessible content.

**Comment by ljharb** [comment]
That's also a good point. Since in the case of disable comments, it'd only require it in the one place, that also seems reasonable.

**Comment by jaepage** [comment]
Note that this very GitHub form itself has two submit buttons ("close and comment" and "comment").

**Comment by jaepage** [comment]
test

**Comment by ljharb** [comment]
That it is done, does not mean it is correct, and the way Github does it is not accessible nor gracefully degradable for the very reasons I've indicated.

**Comment by jessebeach** [comment]
`form`, `table` and `dialog` fall into this weird category where they often host interaction handling. I'm not opposed to adding further configuration to `no-noninteractive-element-interactions` that would whitelist specific handlers on a per-element basis. 

https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/src/index.js#L75

Maybe something like:

```
'jsx-a11y/no-noninteractive-element-interactions': [
  'error',
  {
    handlers: [
      'onClick',
      'onMouseDown',
      'onMouseUp',
      'onKeyPress',
      'onKeyDown',
      'onKeyUp',
    ],
    allowedOnElements: {
        form: ['onKeyPress']
    }
  },
],
```

**Comment by axelboc** [comment]
I'm experiencing this issue with `iframe` and `onLoad`.

**Comment by jessebeach** [comment]
Addressed in #281

**Comment by jessebeach** [comment]
Addressed in #281 .
